#### Importing model - [FinBert](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=3910214)

In [1]:
# Install libraries
!pip install transformers
#!pip install transformers==4.18.0
# Not installing CUDA due to memory issues
!pip3 install torch torchvision torchaudio
!pip install pysentiment2
!pip install textblob

In [2]:
# Import all the necessary packages.
import pandas as pd
import pysentiment2 as ps
from textblob import TextBlob
import numpy as np 
import matplotlib.pyplot as plt
import warnings
import transformers
import torch
from transformers import BertTokenizer, BertForSequenceClassification, pipeline

C:\Users\iwmun\anaconda3\Lib\site-packages\transformers\utils\generic.py:260: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  torch.utils._pytree._register_pytree_node(


In [3]:
print(transformers.__version__)

4.32.1


In [4]:
print(torch.__version__)

2.2.2+cpu


#### Importing text

In [5]:
# Import the data set. We will use the more normalised TextBlob data set. 
speech_finbert = pd.read_csv('speech_bert_1.csv')

As we did with PySentiment2 we are using the TextBlob cleaned dataframe as it is more comprehensive than the Vader cleaned speech data. As well as this we don't need to lemmatise either as the more 'raw' data suits FinBert better

#### Sentiment Analysis

In [6]:
# Load the fine-tuned FinBERT model and tokenizer
model_name = 'yiyanghkust/finbert-tone'
finbert = BertForSequenceClassification.from_pretrained(model_name, num_labels=3)
tokenizer = BertTokenizer.from_pretrained(model_name)

# Set up the sentiment analysis pipeline
nlp = pipeline("text-classification", model=finbert, tokenizer=tokenizer)


In [7]:
# Define a function to apply sentiment analysis
def analyse_sentiment(text):
    result = nlp(text)
    return result[0]['label']

In [8]:
# Apply the sentiment analysis function to the "clean_text" column
speech_finbert['sentiment'] = speech_finbert['clean_text'].apply(analyse_sentiment)

RuntimeError: The size of tensor a (1175) must match the size of tensor b (512) at non-singleton dimension 1

- RuntimeError: The size of tensor a (1175) must match the size of tensor b (512) at non-singleton dimension 1.
- BERT-based models, including FinBERT, have a maximum input length limit of 512 tokens.

In [9]:
# Finding the average amount of tokens in the clean_text column
token_counts = speech_finbert['clean_text'].apply(lambda x: len(tokenizer.tokenize(x)))

average_token_count = token_counts.mean()
print(f"Average number of tokens per cell: {average_token_count}")

Average number of tokens per cell: 2458.701406120761
